### PR TITLE
fix(SF-1168): `mergeSearchNavigationsState` and `mergePastPurchaseNavigationsState` should merge url and store navigations

### PR DESCRIPTION
**Pull request by Vic-Dev**

### CODE REVIEW

File: src/services/urlUtils.ts [hunk 1]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
     };
   };
 
-  export const getAllIds = (request: UrlBeautifier.SearchUrlState) => {
-    return request.refinements.reduce((fields, refinement) => {
-      if (!fields.includes(refinement.field)) {
-        fields.push(refinement.field);
+  // tslint:disable-next-line max-line-length
+  export const getAllIds = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
+    const allIds = state.allIds;
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: Is it okay to be modifying `state`?

File: src/services/urlUtils.ts [hunk 2]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
-      if (field in navigations) {
-        const navigation = navigations[field];
-        navigation.selected.push(navigation.refinements.push(transformed) - 1);
+      if (byId[field]) {
+        const navigation = byId[field];
+        const existingIndex = navigation.refinements
+          .findIndex((ref) =>
+            Adapters.Search.refinementsMatch(<any>transformed, <any>ref, navigation.range ? 'Range' : 'Value')
+          );
+        if (existingIndex !== -1) {
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: Consider inverting the condition and check `existingIndex === 1` first, then do an `else if` clause for the `includes` case to remove a level of nesting.

File: src/services/urlUtils.ts [hunk 3]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
+      byId: getById(state, request),
     };
   };
 
-  export const getAllIds = (request: UrlBeautifier.SearchUrlState) => {
-    return request.refinements.reduce((fields, refinement) => {
-      if (!fields.includes(refinement.field)) {
-        fields.push(refinement.field);
+  // tslint:disable-next-line max-line-length
+  export const getAllIds = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: This change in signature is fine if we're sure nobody else is using it.

File: src/services/urlUtils.ts [hunk 4]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
+    const allIds = state.allIds;
+    request.refinements.forEach((refinement) => {
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: You can `filter` first, then spread the resulting array to `push` if you think that might be clearer.

File: src/services/urlUtils.ts [hunk 5]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
+      if (!allIds.includes(refinement.field)) {
+        allIds.push(refinement.field);
```
Comment (jrmykolyn) [inline src/services/urlUtils.ts]: Does it matter that we're mutating here?

File: src/services/urlUtils.ts [hunk 6]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
-        navigation.selected.push(navigation.refinements.push(transformed) - 1);
+      if (byId[field]) {
+        const navigation = byId[field];
+        const existingIndex = navigation.refinements
+          .findIndex((ref) =>
+            Adapters.Search.refinementsMatch(<any>transformed, <any>ref, navigation.range ? 'Range' : 'Value')
+          );
+        if (existingIndex !== -1) {
+          if (!navigation.selected.includes(existingIndex)) {
+            navigation.selected.push(existingIndex);
```
Comment (jrmykolyn) [inline src/services/urlUtils.ts]: Same as above re: mutation.

File: src/services/urlUtils.ts [hunk 7]
```diff
@@ -122,41 +122,59 @@ namespace UrlUtils {
...
   };
 
-  export const getById = (request: UrlBeautifier.SearchUrlState) => {
-    return request.refinements.reduce((navigations, refinement) => {
+  // tslint:disable-next-line max-line-length
+  export const getById = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
+    const byId = state.byId;
+
+    request.refinements.forEach((refinement) => {
       const field = refinement.field;
```
Comment (jrmykolyn) [inline src/services/urlUtils.ts]: Destructure.

File: src/services/urlUtils.ts (hunk 1 above)
Comment (Vic-Dev) [reply src/services/urlUtils.ts]: no :)

File: src/services/urlUtils.ts (hunk 3 above)
Comment (Vic-Dev) [reply src/services/urlUtils.ts]: we only import the `urlUtils` in `url` and don't export it

Comment (Vic-Dev) [reply src/services/urlUtils.ts]: (also we deleted these functions and I just brought them back)

File: src/services/urlUtils.ts [hunk 8]
```diff
@@ -132,19 +132,16 @@ namespace UrlUtils {
...
   // tslint:disable-next-line max-line-length
   export const getAllIds = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
-    const allIds = state.allIds;
-    request.refinements.forEach((refinement) => {
-      if (!allIds.includes(refinement.field)) {
-        allIds.push(refinement.field);
-      }
-    });
+    const allIds = [...state.allIds];
+    const newIds = request.refinements.filter(({ field }) => !allIds.includes(field)).map(({ field }) => field);
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: I would `map` first, then `filter` so you don't have to destructure twice.

File: src/services/urlUtils.ts [hunk 9]
```diff
@@ -122,41 +122,53 @@ namespace UrlUtils {
...
-    return request.refinements.reduce((fields, refinement) => {
-      if (!fields.includes(refinement.field)) {
-        fields.push(refinement.field);
-      }
-      return fields;
-    }, []);
+  // tslint:disable-next-line max-line-length
+  export const getAllIds = (state: Store.Indexed<Store.Navigation> | Store.AvailableNavigations, request: UrlBeautifier.SearchUrlState) => {
+    const allIds = [...state.allIds];
+    allIds.push(...request.refinements.filter(({ field }) => !allIds.includes(field)).map(({ field }) => field));
```
Comment (jmbtutor) [inline src/services/urlUtils.ts]: I would `map` first then `filter` so you don't destructure twice.

File: src/services/urlUtils.ts (hunk 7 above)
Comment (jrmykolyn) [reply src/services/urlUtils.ts]: Nvm...

File: src/services/urlUtils.ts (hunk 9 above)
Comment (jmbtutor) [reply src/services/urlUtils.ts]: I realized that instead of spreading to `push`, it would be better to just use `concat`.

**Pull request merged**